### ISSUE TITLE
Built in smoother for parameters

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#66]
Could you put in a smoother for parameters?

The only non-obvious thing I would have to do is do some increment of the smoother in the processblock sample loop, easy. Would be pretty easy for you to implement as well.

Oh, and I'd have another place in code where I set the smoothing amount per parameter.

you have getValue, you have setValue. Add getSmoothedValue(), add incSmoothedValue(), or whatever.

I'm building my own system for smoothing for PrettyScope, already did it for Spiral Generator. Just annoying, waste of space in code!

**Comment by elanhickler** [comment]
I wonder if you might want to have a place for a pointer for the smoothing amount value? To avoid making a loop for the smooth parameter slider setter and putting all the parameters I want to smooth in there.

**Comment by elanhickler** [comment]
Oh wait, would this contradict the design of keeping gui and audio separate? Nevermind if it is.

**Comment by RobinSchmidt** [comment]
smoothing - yeah - this would also be nice to have. how does your system work? do you have to re-implement it everytime from scratch? can't you make that reusable and put it somewhere into your shared code section?

the problem i'm seeing with putting that functionality into the Parameter class is to bloat the class up and in some cases, one might not need that feature. of course, i could put it into a subclass SmoothedParameter of Parameter and add the functionality there. ...but then we may also want other features like meta-control, modulation, etc. currently my inheritance hierarchy looks like this:

Parameter < MetaControlledParameter < ModulatableParameter

i might do it like:

Parameter < SmoothedParameter < MetaControlledParameter < ModulatableParameter

...but then i couldn't have non-smoothed meta-controlled parameters. actually, it would be desirable to be able to mix and match these 3 additional functionalities smoothing, meta-control and modulation (and maybe more to come in the future) at will. i'm considering a redesign of the Parameter class using the "Decorator" design pattern:

https://sourcemaking.com/design_patterns/decorator

**Comment by RobinSchmidt** [comment]
not sure, if that's a good idea...

**Comment by elanhickler** [comment]
Right now I my a "myparams" class that holds the information just about everything that I need for conveniently defining parameters, defining callbacks, setting/getting values, and convenience functions to make the code less verbose. (except for maybe positioning on the gui). In myparams class, today I just threw in a rosic::ExponentialSmoother member.

NOTE: For this explanation, not all code is represented, some not-relevant code is removed to make it easier to focus on the content I'm trying to show you. This will be in multiple comments. Next section will explain smoothing.

# :star: Initializing jura::Parameter *
### Here's how I conveniently fill paramStrIds while at the same time defining myparams (calling the constructor) with a cool syntax:
![image](https://user-images.githubusercontent.com/9116575/29956314-fb168754-8e9a-11e7-865e-186e42ae795d.png)

### Now I need to fill each `ptr`, which is a `jura::Parameter *`, for each `myparams`  with a `jura::MetaControlledParameter`:
![image](https://user-images.githubusercontent.com/9116575/29956618-11c2c2c2-8e9d-11e7-86df-801ba55e30d1.png)

#### here's the other usages of paramStrIds
```cpp
// register
void ScopeDisplayOpenGL::connectParameters()
{
	for (auto & s : scope->paramStrIds) 
		s->ptr->registerParameterObserver(this);
}
```
```cpp
// deregister
void ScopeDisplayOpenGL::disconnectParameters()
{
	for (auto & s : scope->paramStrIds) 
		s->ptr->deRegisterParameterObserver(this);
}
```
# :star: Initializing functions for ScopeDisplayOpenGL::parameterChanged
### Before, in `void ScopeDisplayOpenGL::parameterChanged(Parameter* p)` we checked to see what parameter was changed by doing a name lookup. Now I have
![image](https://user-images.githubusercontent.com/9116575/29956981-ba936346-8e9f-11e7-8f1e-531d73e7c128.png)
### inside `ScopeDisplayOpenGL` so I can look it up with a via jura::Parameter *, which I assume is faster than string. I run the function defined inside `initialzeParamLookup`.

```cpp
void ScopeDisplayOpenGL::parameterChanged(Parameter* p)
{/
	if (!initializeParamLookup_was_called)
		initializeParamLookup();

	paramlookup[p](p->getValue());
}
```

### Here's a look at initilizeParamLookup. I set a default callback so we don't have to ask whether a parameter is found or not, because I'm not storing all `jura::Parameter *` in `paramlookup` because only some parameters need a specific function for `ScopeDisplayOpenGL`.
![image](https://user-images.githubusercontent.com/9116575/29956924-34c103e0-8e9f-11e7-87f0-8aa4c308944e.png)

# :star: Initializing widgets
### read comments carefully
```cpp
void PrettyScopeWidgetSection::createWidgets()
{
        /* here I'm just defining some std::functions inside 
         * this function just to keep the code more contained, 
         * and so we can call them later in a loop. */

        /* I had to make 3 different ones because there are 3 kinds
         * of automatable widgets, Button / Slider / ComboBox
         * Robin, there must be a way to avoid this, some kind of base classing 
         * the functions are almost identical...  */

	auto buttonFunc = [this](PrettyScope::myparams & p)
	{
		addWidget(p.button = new jura::AutomatableButton(p.text));
		p.button->assignParameter(p.ptr);
		//p.button->setDescription("");
		p.button->setDescriptionField(infoField);
	};
	auto sliderFunc = [this](PrettyScope::myparams & p)
	{
		addWidget(p.slider = new jura::AutomatableSlider());
		p.slider->assignParameter(p.ptr);
		p.slider->setSliderName(p.text);
		p.slider->setStringConversionFunction(p.stringConvertFunc);
		//p.slider->setDescription("");
		p.slider->setDescriptionField(infoField);
	};

	auto dropdownFunc = [this](PrettyScope::myparams & p)
	{
		addWidget(p.combobox = new jura::AutomatableComboBox());
		p.combobox->assignParameter(p.ptr);
		//p.combobox->setDescription("");
		p.combobox->setDescriptionField(infoField);
		p.combobox->registerComboBoxObserver(this);
	};

        /* Finally, we call the above functions.
         * We check to see what kind of myparam we are looking at.
         * Is it a BUTTON? SLIDER? COMBOBOX?
         * I have an enum for this: enum {BUTTON, SLIDER, COMBOBOX}; */

	for (auto & param : prettyScope->paramStrIds)
	{
		switch (param->type)
		{
		case PrettyScope::BUTTON: buttonFunc(*param); break;
		case PrettyScope::SLIDER: sliderFunc(*param); break;
		case PrettyScope::COMBOBOX: dropdownFunc(*param); break;
		}
	}
}
```

**Comment by elanhickler** [comment]
# :star: myparams Convenience functions explained
### This stuff makes use of `getValue()` `setValue()` * functions from `jura::Parameter *` via boolean signs such as `==`, `!=`, `<`, etc, and most importantly, the equals `=` operator */
```cpp
/* this is inside myparams class */
operator double() const { return ptr->getValue(); }
bool operator==(const double & rhs) const { return ptr->getValue() == rhs; }
bool operator!=(const double & rhs) const { return ptr->getValue() != rhs; }
bool operator<(const double & rhs) const { return ptr->getValue() < rhs; }
bool operator>(const double & rhs) const { return ptr->getValue() > rhs; }
myparams& operator =(double rhs) { ptr->setValue(rhs, true, true); return *this; }
```

### here's some example usage in random places in code of get/set value stuff done with jura::Parameter * (aka ptr from myparams).
```cpp
//set default value in void PrettyScope::createParameters()
LineColorMode = 1.0; 

// checks if FXMode button is on so we can write modified audio to output
if (FXMode) 

//in mosueDrag callback to set ShiftX
scope->ShiftX = +dragValueX + init_ShiftX;
```

# :star: How I did smoothing
### here's what I did inside `myparams` class (remember, it holds EVERYTHING!! HEEHEE!!). I added the smoother object and some smoother convenience functions.
```cpp
/* this is inside myparams class */
rosic::ExponentialSmoother smoother;
void setSmootherTarget(double v) { smoother.setTargetValue(v); }
double getSmootherValue() { return smoother.getSample(); }
```

### now to put it all together
1. spawn a `jura::Parameter *` / `AutomatableSlider *` for `ParameterSmoothing` with the help of `myparams` constructor.
![image](https://user-images.githubusercontent.com/9116575/29957961-311517bc-8ea5-11e7-9f5b-f81ed7441e37.png)

2. impregnate an `std::function` into `ParameterSmoothing`'s `jura::Parameter *` via `setValueChangeCallback()`, put all your `myparams` objects in here that you want to and call their smoothers' `setTimeConstantAndSampleRate`.
![image](https://user-images.githubusercontent.com/9116575/29958029-8a8116f2-8ea5-11e7-8c79-66dd2bc66884.png)

3. for each parameter you want to smooth, shove a little std::function into `jura::Parameter *`. remember, .ptr refers to a `jura::Parameter *` stored in the `myparams` object.
![image](https://user-images.githubusercontent.com/9116575/29959800-ef228296-8ead-11e7-8103-df16e39ce486.png)

4. in our `void PrettyScope::processBlock(double **inOutBuffer, int numChannels, int numSamples)` you then slap this kinda stuff in:
![image](https://user-images.githubusercontent.com/9116575/29959835-11f94944-8eae-11e7-84db-b7c49c748fc5.png)

**Comment by elanhickler** [comment]
# :star: Room for improvement

## Smoothing
### Maybe I would make a myparams manager class instead of using a vector of `myparams *` (aka `paramStrIds`) where I can say which parameters I want to be smoothed. Hmm, but any improvement in this area is probably something up to you, Robin, because, at the moment, I need to explicitly say `setSmootherTarget()` in the `jura::Parameter *` valueChangeCallback.

## Widgets
### In a handful of places in the code I need to do redundant checks to know what kind of widget we're dealing with.

this is inside `myparams` constructor to automatically set default scaling types correctly.
![image](https://user-images.githubusercontent.com/9116575/29963288-1ea0859c-8ebb-11e7-8828-8c515809bd25.png)

this runs the widget instantiation stuff in `createWidgets()`
![image](https://user-images.githubusercontent.com/9116575/29963024-38f17dc6-8eba-11e7-9fe2-3733d7e932d4.png)

convenience function in myparams so I don't have to do, for example, `button->setBounds()` or `slider->setBounds()`. This gets me the right member of `myparams` to then call `setBounds`. I've crashed and been confused a handful of times because I didn't remember to use the right button/slider/combobox here.
![image](https://user-images.githubusercontent.com/9116575/29963061-629689fa-8eba-11e7-9e3b-233ad908d4ba.png)

**Comment by elanhickler** [comment]
ok, just moved myparams to its own class, defined right above PrettyScope class.

**Comment by elanhickler** [comment]
Another thing is that we are smoothing unchanging values. Once the smoothing has reached the target value, we no longer need to continue updating the smoother. If you were to build something into your library, I bet you could account for this, such as: calculate the number of samples it takes to reach target value (such as .000001 delta of current - target), and once `getSample()` has been called that many times, set smoother current value to target value. Not sure if this saves CPU, maybe?

**Comment by elanhickler** [comment]
smoother manager class?
```cpp
// do this somewhere in the code
myparams ShiftX;
ShiftX.smoothingEnabled(true);
ShiftX.setSmoothingAmount(0.3);

// after collecting parameters, do this
for (auto & param : paramStrIds)
	if (param.smoothingEnabled)
		/* Obviously we shouldn't have to create a callback just for this.
		 * Instead, if smoothingEnabled, then just include this as part of whatever
		 * callback is set. However, I don't know how to represent that. That's
		 * up to you, Robin. */
		param.setValueChangedCallback(smootherManager.addParameterForSmoothing(param));

class smootherManager 
{
	someObjectHolder<myparams *> SmoothingList;

	// call this in all the setValueChangedCallbacks
    void addParameterForSmoothing(myparams * p)
    {
		SmoothingList.add(p);
    }

    void removeParameterForSmoothing(myparams * p)
    {
    	SmoothingList.remove(p);
    }

    // call this in the sample block
    void doSmoothing()
	{
		for (auto & param : SmoothingList)
		{
			if (param.isFinishedSmoothing())
				removeParameterForSmoothing(param);
			param.incSmoothedValue();
		}
	}
}

// here's the sample block
whatever sampleBlock(whatever whatever)
{
	smootherManager.doSmoothing();

	double shiftx = ShiftX.getSmoothedValue();
}
```

**Comment by elanhickler** [comment]
here's another version, where myparams does not have any info on smoothers, the smoothers are managed by the smoother manager, and the smoother manager keeps track of what smoother belongs to what myparams

```cpp
class smootherManager 
{
	someObjectHolder<myparams *, ExponentialSmoother> Smoothers;
	someObjectHolder<myparams *> CurSmoothingList;

	void setGlobalSmoothingAmount(double v)
	{
		for (auto & smoother : Smoothers)
			smoother.setSmoothingAmount(v);
	}
	void setSmoothingAmount(myparams * p, double v)
	{
		Smoothers[p].setSmoothingAmount(v);
	}

	void addForSmoothing(myparams * p)
	{
		CurSmoothingList.add(p);
	}
	void removeForSmoothing(myparams * p)
	{
		CurSmoothingList.remove(p);
	}

	void setTargetValue(myparams * p, double v)
	{
		Smoothers[p].setTargetValue(v);
		if (!Smoothers[p].isFinishedSmoothing())
			addForSmoothing(p);
	}
	double getSmoothedValue(myparams * p)
	{
		return Smoothers[p].currentValue();
	}

	// call this in the sample block
	void doSmoothing()
	{
		for (auto & param : CurSmoothingList)
		{
			if (Smoothers[param].isFinishedSmoothing())
				removeForSmoothing(param);
			else
				Smoothers[param].incValue();
		}
	}
}

// here's the sample block
whatever sampleBlock(whatever whatever)
{
	smootherManager.doSmoothing();

	double shiftx = smootherManager.getSmoothedValue(ShiftX);
}

// do this somewhere in the code
myparams ShiftX;
smootherManager.addParameterForSmoothing(ShiftX);
smootherManager.setSmoothingAmount(ShiftX, 0.3);

/* example callback, this should be somehow built in though 
 * otherwise we'd have to write this same thing fo revery
 * parameter we want to smooth. This should be part of the
 * callback if smoothingEnabled is true */
ShiftX.ptr->setValueChangedCallback([this, smootherManager](double v){
	smootherManager.setTargetValue(*this, v);
});
```

**Comment by elanhickler** [comment]
pasted over from the Upsampling/Oversampling for PhaseScope/PrettyScope thread

```cpp
class SmootherManager
{
public:
    SmootherManager() {}
    juce::Array<smoothedParameter *> CurSmoothingList;
    void add(smoothedParameter * sp)
    {
        CurSmoothingList.addIfNotAlreadyThere(sp);
    }

    void remove(smoothedParameter * sp)
    {
        CurSmoothingList.removeFirstMatchingValue(sp);
    }

    void doSmoothing()
    {
        for (auto & sp : CurSmoothingList)
            sp->incValue();
    }
};
```
the object adds itself to the manager, so you need to instantiate the manager and give it a pointer to that so it can add/remove itself.

add itself when value changed
remove itself when value stops changing, or the difference is like... 0.000001. in that case, set the smoother's internal state to the target value, don't leave it at a difference obviously.

**Comment by elanhickler** [comment]
I'm trying to make a parameter smoother system, but it seems like it is going to conflict with the modulation system, because 

right now I have this
```
parTune.setCallback([this](double v) { jbMushroomCore.setPitchOffset(v + parOctave*12); });
```

but I need this:
```
parTune.setCallback([this](double v) { parTune.smoother.setTargetValue(v); });
```

but now the modulation system is going to call `parTune.smoother.setTargetValue()`

dunno what to do.

**Comment by elanhickler** [comment]
It seems that you have to implement a smoother that works alongside the modulation system where changing the UI slider sets the new modulation offset value, that then becomes my "v"

**Comment by elanhickler** [comment]
actually, this is really needed right now. If you aren't going to do it, tell me where in code I can change RS-MET library locally. Here's my code so far if you want to copy and paste it in for yourself:

# ⭐️ ParamSmoother
```cpp
class ParamSmoother
{
public:
	ParamSmoother()
	{
		gaussSmoother.setOrder(8);
		gaussSmoother.setShapeParameter(1);
	}

	void setSampleRate(double v) { sampleRate = v; }

	enum type { EXPONENTIAL, GAUSSIAN, LINEAR };
	void setSmootherType(type v) { smootherType = v; }
	void setSmoothingAmount(double v) { smoothingAmount = v; }

	void inc();

	void setInternalValue(double v);
	void setTargetValue(double v);
	double getValue() { return currentValue; }

protected:
	RAPT::rsSmoothingFilter<double, double> gaussSmoother;
	rosic::ExponentialSmoother expoSmoother;
	double sampleRate;
	int smootherType = 0;
	double currentValue, targetValue;
	double smoothingAmount = 0.0;
	bool isPaused = false;
};

void ParamSmoother::inc()
{
	if (isPaused)
		return;

	switch (smootherType)
	{
	case EXPONENTIAL:
		currentValue = expoSmoother.getSample();
		break;
	case GAUSSIAN:
		currentValue = gaussSmoother.getSample(targetValue);
		break;
	}
}

void ParamSmoother::setTargetValue(double v)
{
	targetValue = v;
	switch (smootherType)
	{
	case EXPONENTIAL:
		expoSmoother.setTargetValue(v);
		break;
	case GAUSSIAN:
		break;
	}
}


void ParamSmoother::setInternalValue(double v)
{
	currentValue = v;
	switch (smootherType)
	{
	case EXPONENTIAL:
		expoSmoother.setCurrentValue(v);
		break;
	case GAUSSIAN:
		gaussSmoother.setTimeConstantAndSampleRate(0, sampleRate);
		gaussSmoother.getSample(v);
		gaussSmoother.setTimeConstantAndSampleRate(smoothingAmount, sampleRate);
		break;
	}
}
```

# ⭐️ ParamManager
```cpp
class ParamManager
{
public:
	ParamManager() {}

	ParamManager(vector<myparams*> ParamList) : ParamList(ParamList) 
	{
		for (auto & p : ParamList)
			if (p->shouldBeSmoothed)
				paramsThatShouldBeSmoothed.push_back(p);
	}

	void setSampleRate(double v)
	{
		for (auto & p : paramsThatShouldBeSmoothed)
			p->smoother.setSampleRate(v);
	}

	void addForSmoothing(myparams * sp)
	{
		paramsToSmooth.addIfNotAlreadyThere(sp);
	}

	void removeForSmoothing(myparams * sp)
	{
		paramsToSmooth.removeFirstMatchingValue(sp);
	}
	
	void doSmoothing()
	{
		for (auto & p : paramsToSmooth)
			p->smoother.inc();

	}

	int getNumParamsToSmooth()
	{
		return paramsToSmooth.size();
	}

protected:
	vector<myparams*> ParamList;
	juce::Array<myparams*> paramsToSmooth;
	vector<myparams*> paramsThatShouldBeSmoothed;
};
```

**Comment by elanhickler** [comment]
I found ModulationTarget class with this:

```cpp
  void setUnmodulatedValue(double newValue)
  {
    unmodulatedValue = newValue;
  }
```

So all that is needed is to call `setUnmodulatedValue()` for slider changes based on the smoothed value or something.

**Comment by elanhickler** [comment]
If you don't quickly make a smoothing system, maybe you could quickly separate out the modulation and slider change callbacks so I can do:

```cpp
parTune.setCallback([this](double v) 
{
    //ptr is jura::ModulatableParameter2
    parTune.modValue = parTune.ptr->getModulationValue();
    parTune.sliderValue = parTune.ptr->getUnmodulatedValue();
    parTune.smoother.setTargetValue(sliderValue);
    // in my per sample callback I do parTune.smoother.inc();
    jbMushroomCore.setPitchOffset(parTune.smoother.getValue()); 
});
```

Boom! Done.

**Comment by elanhickler** [comment]
I need a sliderChangeCallback, so I can add parameters to my smoother manager only when the slider changes rather than when the modulation changes

**Comment by elanhickler** [comment]
ok, i finished implementing a smoother manager to my satisfaction, you no longer need to rush this. Implement it whenever.

**Comment by elanhickler** [comment]
fuck. nope. Smoothing doesn't work because the modulated is inexplicably tied to to the slider value.

If there's no modulator, my smoother setup works. If there's a modulator, than the modulation bias is based on the slider value or something. I need smoothing! Do something like give me some callbacks to work inside or some functions to retrieve the raw modulation value before it gets affected by the slider or something!!!!

...I guess i can release some free products with this smoothing issue.

**Comment by RobinSchmidt** [comment]
is you smoother calling ModulatableParameter::setValue per sample with the smoothed (lowpassed) value? if, so - what kind of problem are you experiencing? actually, this call would set up the unmodulatedValue in the ModulationTarget class, which in turn would get modulated by the modulators. and if the unmodulated value changes smoothly, so should the modulated value as well. the modulated value should just smoothly follow the unmodulated one, if i'm not mistaken.

**Comment by RobinSchmidt** [comment]
i could certainly add an getUnmodulatedValue function to ModulationTarget, if you need to inquire that value. but i would think, what you actually need is a setter - and such a setter is already there. ModulatableParameter::setValue calls ModulationTarget::setUnmodulatedValue (which you could also call directly, if deisred, but i think ModulatableParameter::setValue is appropriate)

**Comment by RobinSchmidt** [comment]
>  Do something like give me some callbacks to work inside or some functions to retrieve the raw modulation value before it gets affected by the slider

actually, there is no such thing as a "raw modulation value" that "gets affected by the slider". is the other way around: there's a raw (unmodulated) value that is set up by the slider and *that* value gets affected by the modulators

**Comment by RobinSchmidt** [comment]
so if you smoothly change that nominal unmodulated reference value, it should actually work...unless i'm missing something

**Comment by RobinSchmidt** [comment]
> ModulatableParameter::setValue calls ModulationTarget::setUnmodulatedValue (which you could also call directly, if deisred, but i think ModulatableParameter::setValue is appropriate)

wait - no - you should probably better call ModulationTarget::setUnmodulatedValue because ModulatableParameter::setValue also calls MetaControlledParameter::setValue - and that notifies the host about plugin automatable parameter changes. you probably don't want that to happen per sample

**Comment by elanhickler** [comment]
robin, just tell me what to do! Don't add features unless what I want to do is impossible. Ok, let me try what you are suggesting. `ModulationTarget::setUnmodulatedValue`

**Comment by RobinSchmidt** [comment]
i don't know your smoothing system well enough to be exactly sure what to do. but my suggestion would be to call ModulationTarget::setUnmodulatedValue from your per-sample update function in your smoother. i think, that should work. like per sample call 

`myParam->setUnmodulatedValue(mySmoothingLowpass.getSample(targetValue))`

**Comment by elanhickler** [comment]
ok, I can set the unmodulated value, sure, but moving the slider still sets the value too. So how do I stop that and only have the value be set by the smoother?

**Comment by elanhickler** [comment]
Right now this happens.

1. I move a slider.
2. Parameter callback is called with input value of slider.
3. Ignore that and do getValue() to get value of just slider without modulation.
4. By the way, modulators call this callback when they update value, too. Ignore that and take modulated value into account before I set the value of the underlying object with getModulatedValue(),
5. Finally, whenever I increment the smoothers, I do the "callCallbacks" function to update the underlying object, all the while modulation updates will also call the callback, thereby executing all my smoothing code, but i have if statements to bypass when smoothing is already happening.

I think just I need a separate callback for slider updates.

**Comment by elanhickler** [comment]
I need:

1. move slider
2. do slider changed callback
3. capture slider value
4. set smoother target value
5. increment smoother
6. smoother sets unmodulated value
6. when modulation happens, that calls the value change callback setting the modulated value
7. FINALLY: send unmodulated value + modulated value to underlying object

**Comment by elanhickler** [comment]
My smoother manager code solid, all you would have to do is copy/paste that into handling the underlying slider value.

# .h

```
class SmootherManager
{
	friend class myparams;
public:
	SmootherManager() {}
	// add available parameters that need smoothing
	void addParameters(vector<myparams*> list);
	// sets samplerate for all smoothers
	void setSampleRate(double v);
	// call this per sample to do smothing
	void doSmoothing();
	// call this in block AFTER smoothing to remove "finished" smoothers
	void cleanup();
	// call is in block to know if smoothing should take place per sample
	int getNumParamsToSmooth();
	// convenience function to set smoothing amount for all smoothers
	void setGlobalSmoothingAmount(double v);

protected:
	void addForSmoothing(myparams * sp);
	void removeForSmoothing(myparams * sp);	
	vector<myparams*> ParamList;
	juce::Array<myparams*> paramsToSmooth;
	vector<myparams*> paramsThatShouldBeSmoothed;
};
```
#.cpp
```
void ParamManager::addParameters(vector<myparams*> list)
{
	ParamList.insert(ParamList.end(), list.begin(), list.end());

	int i = 0; // robin doesn't need this

	for (auto & p : ParamList)
	{
		p->id = i++; // robin doesn't need this
		p->managerPtr = this;

		// robin doesn't need this
		// this is to be more efficient for setSampleRate and setGlobalSmoothingAmount
		// we have a master list of parameters and from that list we get
		// another list just for those parameters that want smoothing
		if (p->shouldBeSmoothed) 
			paramsThatShouldBeSmoothed.push_back(p);
	}
}

void ParamManager::setSampleRate(double v)
{
	for (auto & p : paramsThatShouldBeSmoothed)
		p->smoother.setSampleRate(v);
}

void ParamManager::setGlobalSmoothingAmount(double v)
{
	for (auto & p : paramsThatShouldBeSmoothed)
		p->smoother.setSmoothingAmount(v);
}

// robin, you could have the underlying parameter
// add itself and keep track of if it was added
// rather than the manager. I did this this way because
// it's cleaner code for dealing with the lack
// of proper callbacks.
void ParamManager::addForSmoothing(myparams * p)
{
	if (p->isBeingSmoothedByManager())
		return;

	paramsToSmooth.add(p);

	p->isAddedToSmootherManager = true;
}

// again, you can have the underlying parameter
// do this stuff.
void ParamManager::removeForSmoothing(myparams * p)
{
	paramsToSmooth.removeFirstMatchingValue(p);
	p->isAddedToSmootherManager = false;
}

// not sure how else to do it with
// the lack of proper callbacks
void ParamManager::doSmoothing()
{
	for (auto & p : paramsToSmooth)
	{
		p->smoother.inc();		
		p->ptr->callValueChangeCallbacks();
	}
}

void ParamManager::cleanup()
{
	for (auto & p : paramsToSmooth)
		if (!p->smoother.needsSmoothing())
			removeForSmoothing(p);
}

int ParamManager::getNumParamsToSmooth()
{
	return paramsToSmooth.size();
}
```

**Comment by elanhickler** [comment]
Here I'm pulling some code for the actually smoother class just so you see how I make sure we don't smooth parameters that no longer need smoothing.

```cpp
// every sample we check to see if we are close enough to target
// if so, call end smoothing so that needs_smoothing becomes false
void ParamSmoother::inc()
{
	if (!needs_smoothing || isPaused)
		return;

	currentValue = expoSmoother.getSample();	

	if (fabs(currentValue - targetValue) <= 1.e-6)
		endSmoothing();
}

// when we set a new target, set needs_smoothing to true
// you could also check to see if smoothing speed is
// negligible and just call endSmoothing if so and
// keep needs_smoothing false
void ParamSmoother::setTargetValue(double v)
{
	targetValue = v;
	expoSmoother.setTargetValue(v);
	needs_smoothing = true;
}

void ParamSmoother::endSmoothing()
{
	currentValue = targetValue;
	expoSmoother.setInternalValue(targetValue);
	needs_smoothing = false;
}

void ParamSmoother::getCurrentValue()
{
	return currentValue;
}
```

**Comment by elanhickler** [comment]
Robin, I just realized I need a smoother manager for my synth for general purposes of smoothing things like pitch and just random stuff.

If you create a built in smoother for parameters, I'm guessing I can't also use it to smooth arbitrary variables in my synth. So I should keep my smoother manager? I could repurpose it just for arbitrary variables not having to do with parameters.

**Comment by RobinSchmidt** [comment]
oh - for arbitrary variables. that's an interesting design challenge. how would you approach that? i guess, it would have to based on some member-function pointer where the member function to be called must adhere to a given signature, such as: `void setSomething(double something);` ....or something. in a similar way, the valueChangeCallback of the Parameter class works. unless you want to force your dsp-classes with smoothable parameters to be subclasses of some baseclass

**Comment by elanhickler** [comment]
To make things uncomplicated, and also allow the smoother manager to do more than just parameters, the class that needs a manager has a smoother manager point AND smoother objects, for example

```cpp
what about something like this

```cpp
class SmootherManager
{
public:
	Array<Smoother *> smoothersCurrentlyActive;

	void addForSmoothing(Smoother * smoother)
	{
		if (!smoother->wasAlreadyAdded)
		{
			smoothersCurrentlyActive.push(smoother);
			smoother->wasAlreadyAdded = true; // set to false before removing smoother
			// manager removes smoother when it's done smoothing
		}
	}	
};

class Monosynth
{
public:
	SmootherManager * smootherManagr;

	Smoother pitchSmoother;

	void frequencyChanged()
	{
		smootherManager->addForSmoothing(pitchSmoother);
	}

	updateFrequency()
	{

	}

	void processPerSample()
	{
		if (pitchSmoother.valueHasChanged())
			updateFrequency();
	}
};
```

**Comment by RobinSchmidt** [comment]
do you want for every smoothable parameter a paramaterChanged and updateParameter function? ....and a big series if if(thisSmoother.valueHasChanged)... i think, this will lead to a proliferation of boilerplate code real quick

**Comment by RobinSchmidt** [comment]
also, your monosynth is coupled with the smoother. i'd prefer a design that decouples the monosynth from the smoother/manager

**Comment by elanhickler** [comment]
ok, maybe you have better ideas. I wouldn't worry about supporting arbitrary variables unless you see an easy way to do it.

**Comment by RobinSchmidt** [comment]
i would probably attempt something based on a callback, like in Parameter. a smoother, gets a target value, (and currentl value) and an member-function-callback-object. ...and then the smoother manager would call some per samples update-function on each of it's smoothers. the target class, i.e. the smoothee, would not have to know about the existence of the smoother class at all. at could be, for example, a filter - and the smoother would then just call setCutoff on it, per sample, via the callback-object

**Comment by RobinSchmidt** [comment]
....but: how to marry that with the mod-system? that adds another layer of complication

**Comment by elanhickler** [comment]
That seems like the easy part. Your mod system has the "unmodified value", that's the one you smooth/update. You know my smoother manager works with your mod system, it only doesn't work for child objects because I didn't provide them with a smoother manager pointer, I'd need to rewrite some code for that.

**Comment by elanhickler** [comment]
oh, well you're talking about value change callbacks, nevermind. In my value change callback I do `smoother.getSmoothedValue()` essentially.

**Comment by RobinSchmidt** [comment]
hmm...well, ok, yes - you could give the smoother the ModulatableParameter as target "smothee" object with the setUnmodulatedValue as target callback. that should probably work

**Comment by RobinSchmidt** [comment]
did your smoother actually work well together with the meta-system? such that, for example, when 2 parameters were assigned to the same meta and you moved one, both were smoothed? where did you set up your target-value? in the slider callback? because i think, in this case, it should not work out with the meta-system. i think, to make it all work, i'll need to change my inheritance hierarchy to something like:

Parameter < Modulated < Smoothed < MetaControlled

...but where would then later a Poly-Parameter go? maybe between Modulated and Smoothed....but then we would always have the data-overhead of mod and poly, even for parameters where just smoothing and meta-control is needed

**Comment by RobinSchmidt** [comment]
because when a meta changes, it should set the target of the smoother (so "smoothed" must be inside "meta"). the smoother in turn should change the unmodulatedValue of the modulated parameter (so "modulated" must be inside "smoothed")

**Comment by elanhickler** [comment]
Smoother was fine with meta-parameter system.

**Comment by RobinSchmidt** [comment]
really? that would sort of surprise me. can you try to assign some arbitrary parameter and DC to the same meta, move the "arbitrary" slider and see if you get smoothed dc?

**Comment by elanhickler** [comment]
works fine. I still get clicks due to having your modulation stuff in my smoothing code. But i think we can safely assume it is it working as intended.

**Comment by RobinSchmidt** [comment]
...i mean, if you set the smoother target in the slider callback, how would that be supposed to work?

**Comment by RobinSchmidt** [comment]
hmm...ok..

**Comment by elanhickler** [comment]
![image](https://user-images.githubusercontent.com/9116575/32394742-610d600a-c09b-11e7-81d3-5344d9b887ff.png)

not sure what you are using this for, but you can store a bool "isAddedToWhatever" inside the object that you are adding to an array.

addObjectToAray(object * obj)
{
    obj->wasAdded = true;
    myArray.add(object); // does not require a search
}

removeObjectFromArray(object * obj)
{
    obj->wasAdded = false;
    myArray.remove(object); // this may require a search
}

**Comment by RobinSchmidt** [comment]
i'm using this for general purpose adding of an object to a std::vector, if it's not already inside the vector. "general purpose" here means that it's not feasible in general to store a "wasAdded" flag inside the object. you could do this thing with special purpose arrays - but it would imply that the to-be-stored objects are subclasses of some baseclass that has a "wasAdded" flag. so you would have a coupling between the array container class and the element class

**Comment by elanhickler** [comment]
> ...also what would you do, if the object is possibly element of several arrays at the same time?
just don't 👍

**Comment by RobinSchmidt** [comment]
for example, my ModulationManager keeps an array of availableTargets and affectedTargets and it totally makes sense.

**Comment by elanhickler** [comment]
> although, when i look at mushroom, i'm not so sure anymore about the validity of this assumption).

I want to make it more modular where you can add/remove LFOs/ADSRs/Breakpoints from the GUI, so the available sources may start off as just a few until the user adds stuff. I want to go for a wireless modular synth.

**Comment by elanhickler** [comment]
please put a pause function on a smoother. I need to be able to pause the value from changing.

or I could just set smoothing speed to infinity for specific parameters when needed.

**Comment by RobinSchmidt** [comment]
oh! what is that good for?

**Comment by elanhickler** [comment]
for pausing PrettyScope. You click on the canvas to draw something, the x/y offset starts moving, you release mouse, the x/y needs to stop moving. This is "mouse auto pause mode" so it makes it easier to draw stuff.

**Comment by RobinSchmidt** [comment]
so, you mean, you need it to stop immediately on releasing the mouse instead of keep moving and slowing down for some further fractions of a second?

**Comment by elanhickler** [comment]
huh? "instead of moving and slowing down" ? yeah, it has to stop changing value... but that's for me to worry about, all I need is to be able to change the speed of the smoother.

**Comment by RobinSchmidt** [comment]
yes - actually, per smoothing-target and the smoother then takes over the value from the target when the target gets assigned. but that's technical details

**Comment by RobinSchmidt** [comment]
okayyyy....the smoother is operational. i tested it in Chainer with FuncShaper. you can try it with the dc-offset value. i also tested to assign it to a meta (and drive to the same meta) and that seems to work as well. modulation also. 

you need to create your parameters as SmoothableParameter now. it's a subclass of ModulatableParameter. so currently i have the rather weird inheritance hierarchy:

Parameter < MetaControlledParameter < ModulatableParameter < SmoothableParameter

although it seems to work, i want to try to change it into:

Parameter < SmoothableParameter < MetaControlledParameter < ModulatableParameter 

we'll see how that works out. when i'm done with that, you can change your code back to creating ModulatableParameter objects. but if you want to try and use it *right now*, create SmoothableParameters

**Comment by RobinSchmidt** [comment]
btw. it will also get invoked when you switch presets. i guess, when setting up a rather long smoothing time, we'll get a preset-morph when switching presets. ...but that will almost certainly create a huge cpu spike. wen may consider to temporarily turn smoothing off on preset switch. ...i need to include some facilities for that...

**Comment by RobinSchmidt** [comment]
oh - wait - now it's not working anymore...wtf?

**Comment by elanhickler** [comment]
Yes, I want to use it immediately, I don't care if I have to change code later, I need to release Mushroom Generator ASAP.

But looks like I'll need to wait until you solve the meta attach issue.

**Comment by RobinSchmidt** [comment]
i think, i first need to change my inheritance hierarchy and see how it will go then.

**Comment by RobinSchmidt** [comment]
ok, it seems to work. you can keep your code as is - ModulatableParameter is now the outermost class and includes smoothing and meta-control. but for simpler plugins that don't use the mod-system, you can just create MetaControlledParameters to avoid the mod-system overhead

**Comment by RobinSchmidt** [comment]
oh - i have no thread-safety measures in place yet. ...will add that next

**Comment by elanhickler** [comment]
quick description of how to use?

**Comment by elanhickler** [comment]
do I need to create my own smoothing manager like for modulation?

**Comment by RobinSchmidt** [comment]
> quick description of how to use?

just look at processBlock of AudioModuleChain (that's the awkward class name of the Chainer). it's very similar to your smoother

> And aren't you excited to employ the smoothing manager for morphable presets?

it might be fun to noodle around with but ultimately probably impractical for release due to the high cpu-load during the morph

> How do I turn off smoothing for certain sliders? edit: oh I guess i can put smoothing to 0

exactly

> Smoother Manager needs a "set smoothing amount for all sliders" function

hmm...ok, i think, i can add an option to override the individual setting in the smoothing-target

> do I need to create my own smoothing manager like for modulation?

what? why?

**Comment by elanhickler** [comment]
thinking about it, if I need to disable smoothing and implement my own pause functionality, I won't use the smoother manager's "set all sliders speed".

**Comment by RobinSchmidt** [comment]
i could perhaps provide a global switch for turning smoothing off. might be needed anyway to disable it temporarily during preset recall for reasons stated above. ...but yeah, if you need some very special functionality that is relevant only to PrettyScope, it's probably best to do it in a subclass

**Comment by elanhickler** [comment]
pause functionality... preset morphing... disabling smoother for certain parameters... I need to think more about all of this... let's start a preset morphing discussion for fun.

For now, please get on adding grid settings saving to breakpoint module.

Once that's done, I still need work on the phaselocking autotuner.

I also need you to perhaps work on the multiband compressor, I'll fund you with the money coming from that person / project.

There's a lot to do still!!!!!!

**Comment by RobinSchmidt** [comment]
....perhaps the smoother could have an undersampling function to get the cpu load down to managable levels, such that it doesn't update parameters every sample but only every N samples, where N is a user-adjustable undersampling factor.

**Comment by elanhickler** [comment]
undersampling... that reminds me. I'm also doing some SERIOUS code refactoring for prettyscope, I very badly want an upsampling method for it, it would improve the visuals by a factor of 10.

**Comment by RobinSchmidt** [comment]
hmm...yeah - as said, in PhaseScope, i would probably try to replace the connecting lines with cubic splines but such a thing is not so easy to realize in PrettyScope, because OpenGL has no spline primitive, so at the end, the splines would have to be approximated by a sequence of lines anyway. within my own c++ pixel-code, i could add such a spline-drawing primitive

**Comment by elanhickler** [comment]
right, wouldn't we just upsample the audio buffer? the audio buffer is what creates the x/y points.

**Comment by RobinSchmidt** [comment]
you could do that - but i'd certainly not recommend using an elliptic filter for the upsampling (unless you like to see spurious gibbs ripples in the plot). you want something with good time-domain performance, i.e. bessel or maybe gaussian. the upsampled signal would not necessarily go through the sample-points than anymore, though. with splines, it would.

**Comment by elanhickler** [comment]
I don't understand the issue. So.............. what needs to be done? Do we need an openGL expert to make splines instead of lines?

**Comment by elanhickler** [comment]
This is why I posted dood.al source code.

**Comment by RobinSchmidt** [comment]
does that use splines?

**Comment by elanhickler** [comment]
ok i made some section links for easy navigation

https://gitlab.com/Hickler/Soundemote/issues/58

**Comment by elanhickler** [comment]
My value change callbacks are not being hit. It's not even a problem of smoothing. Just simply moving a slider doesn't trigger a callback. My breakpoint isn't being hit for that callback.

![image](https://user-images.githubusercontent.com/9116575/32466294-af4f059c-c303-11e7-8bb5-867526867818.png)

![image](https://user-images.githubusercontent.com/9116575/32466309-baaec828-c303-11e7-9357-824b5765af6e.png)

![image](https://user-images.githubusercontent.com/9116575/32466319-c5084510-c303-11e7-988e-5fbc0603ac98.png)

**Comment by elanhickler** [comment]
![image](https://user-images.githubusercontent.com/9116575/32466748-412b90e2-c305-11e7-9624-7a60203f0e94.png)

call value change callbacks is false when i move a slider???

**Comment by RobinSchmidt** [comment]
hmm...i guess, you are coming from here?:
```cpp
void rsSmoothableParameter::setValue(double newValue, bool sendNotification, bool callCallbacks)
{
  if(smoothingTime == 0.0 || smoothingManager == nullptr)
    Parameter::setValue(newValue, sendNotification, callCallbacks);
  else
  {
    double oldValue = getValue();
    Parameter::setValue(newValue, sendNotification, false);
    smoothingManager->addSmootherFor(this, newValue, oldValue);
  }
}
```
yes, when the baseclass method is invoked, i pass false to the "callCallbacks" parameter because the callback call is deferred to the "updatedSmoothedValues" call which is supposed to be called right before the next sample is supposed to be produced

**Comment by elanhickler** [comment]
yes
![image](https://user-images.githubusercontent.com/9116575/32467258-2e80ba4c-c307-11e7-8022-b14bd9128792.png)


so why no work?

**Comment by RobinSchmidt** [comment]
i'll have a look at it in the debugger myself tomorrow

**Comment by elanhickler** [comment]
did you forget that I am using modulatable parameter 2 for std::function?

**Comment by elanhickler** [comment]
Assigning modulation to a parameter, even if that modulation amount is 0, will allow slider to change value, however it is not smoothed.

`smoothingManager.needsSmoothing();` is never true

**Comment by RobinSchmidt** [comment]
ok, i fixed some bug. you created your own smoothing manager, but that's not how it's supposed to work. the smoothing manager exists one in the AudioPlugin object and is accessible globally by all AudioModules via a pointer (defined in baseclas AudioModule). ...so now the parameter setvalue functions get called...but it seems not being smoothed. but maybe that's now easier for you to figure out

**Comment by elanhickler** [comment]
no idea why parameters are not smoothed.

what else do you have to do to make smoothing work?

im lost.

**Comment by elanhickler** [comment]
is smoothing speed 0 by default?

**Comment by RobinSchmidt** [comment]
oh - i see in the debugger that smoothingTime is set to 0.01. but that's in milliseconds!

**Comment by elanhickler** [comment]
If you're dividing smoothing time by 1000, that's kinda pointless.

**Comment by RobinSchmidt** [comment]
hmm...it's a matter of convention what units should be used for the parameters - seconds vs milliseconds in this case. i thought, milliseconds would be the more convenient unit here

**Comment by elanhickler** [comment]
I don't expect you to change this, but just think about it. Keep all time units in seconds throughout your library. To have multiple units of time is just an extra layer of complexity.

For example, I can make a slider display milliseconds or seconds, but the underlying value is always in seconds. Typing 1 vs .001 in code... well you don't know by looking at it whether it's a second of a millisecond or a millionth of a second.

Keeping time units the same means you never have to remember to multiply by 1000 or not for this or that object or parameter.

**Comment by elanhickler** [comment]
If you want milliseconds then... you should name your function setSmoothingTimeMilliseconds... which will then make you not want to set things in milliseconds because of the extra long function name 😄

**Comment by RobinSchmidt** [comment]
hmmm....yeaaa....i sort of agree that consistency of units throughout a library is desirable. i was certainly thinking about that for rapt. in the rosic classes, i mostly use units of whatever i considered the most convenient unit to present to an end user on the gui - like milliseconds for compressors but seconds for reverb-times and so on. but in the jura framework i'm a bit more sloppy because i consider it mainly for personal use anyway. hmmm

**Comment by elanhickler** [comment]
Robin I wish I could just write all my own code and not use your library so you wouldn't have to worry about me and my opinions, but........... I'm here to make money and I can't afford to write my own stuff due to time and lack of knowledge. If my business is successful, more money will be flowing your way. 

If it were up to me, I would continue to pay for you to develop your interface with public use in mind and create an awesome one-stop framework to build a plugin from scratch.... well that's what you're doing right now! There's money potential here. You're doing it anyway, so why keep saying it's for personal use? You can keep saying that, but I think there's a lot of potential here for your business to grow.

I could be wrong.

**Comment by RobinSchmidt** [comment]
i'm shying away from advertising my framework it as a complete one-stop solution for building plugins because i feel a bit shaky about how good it is from a software engineering point of view. i mean - damn - i'm using mutex locks in realtime code! ...for example. (although, from what i've heard, iplug/wdl does that too - but the juce guys would certainly sneer at that). i feel much more confident about the value of my math and dsp code, though

**Comment by RobinSchmidt** [comment]
actually, juce itself is already supposed to *be* that framework

**Comment by elanhickler** [comment]
i get that, and I understand, but for me as a "not programmer" I don't care, as long as it works. You can fix that stuff over time.

Any time someone wants to build a plugin I would tell them to use your library for everything. Or that's what I want to do. If they are confused as to how to build a plugin, they are not experienced enough to care about those details.

Also, no idea what mutex locks are.

**Comment by elanhickler** [comment]
juce is supposed to be the framework sure

but what about modulation manager, smoothing manager, metaparameter manager? that's what I can't just easily replace.

**Comment by RobinSchmidt** [comment]
yes, and not to forget, to (almost) totally delegate away state-recall to framework code

**Comment by elanhickler** [comment]
but if only noobie programmers are using your library that may not do much for your business, you'd need tons of noobie programmers, like an alternative to max/msp

**Comment by elanhickler** [comment]
This topic can be closed!

**Comment by RobinSchmidt** [comment]
...so does the smoother work for you now? but i think, a few details still need to be refined - like the initialization and preset-load issues.

> no idea what mutex locks are.

a rather brute force way of ensuring thread-safety. the more sophisticated way would be to use a lockfree queue. ...which is quite a bit more complicated to implement

**Comment by elanhickler** [comment]
you once tried to make money with plugins, that means something.

Why not let me make plugins for you? I have 3 or even 4 potential customers which want plugins.

Jamal, the multiband compressor project, Greg from Orange Tree Samples, maybe Embertone (although they want iOS apps).

The only reason those projects are not moving forward is because I'm waiting on you for certain things.

**Comment by RobinSchmidt** [comment]
> Why not let me make plugins for you?

you can do that. i would then charge for the dsp library and possibly write some more custom dsp code for the project in question. ok, maybe framework code, too

> I'm waiting on you for certain things.

yes, but the gaps are filling. we have a working meta-system, mod-system and smoothing-system now. i guess next would be the meta-mapper (which i'd do free of charge, as said) and gui customization?

**Comment by elanhickler** [comment]
I want to release ~$200 worth of plugins by the end of the year.

If I can continue to commission you to program stuff so that I don't have to, that is what I want to do, because my time is needed to run my business in other ways. For example, I'd comission you to create an X/Y widget, or different gui elements, or a preset browser, etc. There's a lot of things I would pay for.

Forget the meta-mapper. I need to think about that some more and there's a lot I can do with what we have right now.

**Comment by elanhickler** [comment]
OH wait, don't forget the meta-mapper... your plan is to more or less create a "breakpoint envelope thing" except that a knobs can control where the "playhead" is positioned within that envelope. Then I can use the modulation manager to route things however.

**Comment by RobinSchmidt** [comment]
> your plan is to more or less create a "breakpoint envelope thing" except that a knobs can control where the "playhead" is positioned within that envelope. 

yes

> Then I can use the modulation manager to route things however.

actually i was more thinking about using this to define the mapping function for meta-parameters to their dependent parameters. not sure how to use it with the mod-system, but perhaps possible, too. the core-mapper itself would just take an input value and spit out a corresponding output value - so it could be used in various contexts. in fact, i already have something quite similar in rosic (class PiecewiseFunction)

**Comment by elanhickler** [comment]
not sure how to use it with mod system? easy.

Just like I instantiate breakpoint envelope generators, and send those to sliders... ok this would be a little complicated, but fine for something quick and dirty, not that I need quick and dirty, I don't need it right now anyway, but...

breakpoint --> slider B, slider C, slider D (using modulation manager)

slider A --> breakpoint position (using meta manager)

**Comment by RobinSchmidt** [comment]
sliders for breakpoints? i was thinking more about a gui editor similar to the breakpoint modulator editor - where you insert/remove/move-around nodes. in fact, i'm thinking about making a general node-editor baseclass - i have a lot of plot-editors where nodes are moved around (echolab, easyq, etc). it could make sense to factor out some things that are common to all these editors

**Comment by elanhickler** [comment]
quick question, is there a way to access to child module jura::ModulatableParameters so i can set their smoothing speed?

**Comment by RobinSchmidt** [comment]
so, you want to set the smoothing speed for all parameters in an AudioModule (and also in its child-modules, recursively) at once? there's no convenient way for this yet, but could easily be added, i think.

**Comment by elanhickler** [comment]
nooo, see, you can't add that function, because then it's going to set parameters that should not be smoothed, that should be kept at a speed of 0.

The way I did it with my smoother is by keeping a list of parameters that should be smoothed and only iterating through those for any smoother functions. But for your smoother you don't have that differentiation.

Anyway, I just made a way to do it for myself.

You can think about adding a few more conveniences to your smoother whenever.

**Comment by elanhickler** [comment]
And eventually, different sliders will need different smoothing speeds. I can already tell one global smoothing about isn't correct, but that's a fine detail about parameters I have no intention to tackle until many other things are solved.

**Comment by elanhickler** [comment]
going to close this, let's continue related discussions in new/old threads.

**Comment by RobinSchmidt** [comment]
>  you can't add that function, because then it's going to set parameters that should not be smoothed, that should be kept at a speed of 0.

hmm - are they to be meta-controlled and/or modulated or are they just static settings? because in this case, you could just create them as objects of the Parameter baseclass

**Comment by elanhickler** [comment]
yes they may be modulated and meta-controlled and all of that.

For example, octave slider is integer, does not need to be smoothed, but definitely still want to modulate and automate it.

Actually, anything that has integer values should not be smoothed.

**Comment by RobinSchmidt** [comment]
ok...i seee...i guess i could add a flag to the rsSmoothingTarget class to indicate whether or not it should be smoothed

**Comment by elanhickler** [comment]
UNLESS you want to do something more fancy with integer sliders:

slide to integer values or whatever the division.

since I actually have smoothing on my octave parameter, but to achieve that I have to quantize the incoming value, then smooth that by a desired amount.

**Comment by elanhickler** [comment]
UNLESS you want to do something more fancy * 2

allow sliders to be optionally quantized via right click menu. heehee.

But no don't do that, I have some very specific things in mind for such functionality, it needs a dedicated gui rather than a right click menu.
